proposed/2022/version-overrides.md

```diff
@@ -0,0 +1,119 @@
+
+- A project exists in a repository but is not part of the main set of shipping assemblies and is allowed to use newer, untested package depedencies (such as an experimental tool)
+- A project that has more advanced requirements and cannot use a single version of a package dependency for all of its target frameworks
+
+In these cases, a project can override the centrally defined package version.  However, if that project is part of a transitive graph with version conflicts, a user would still need to resolve those conflicts before restore would succeed.  The main point of this feature is to give a well documented and supported way of overriding a package version for a particular project, not to make graph resolution easier if package conflicts exist.  
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+
+First a user opts-in to Central Package Management with a `Directory.Packages.props` file in their directory tree.
+
+A user would specify package versions in `Directory.Packages.props`:
+
+```xml
+<Project>
+  <ItemGroup>
+    <PackageVersion Include="PackageA" Version="1.0.0" />
+    <PackageVersion Include="PackageB" Version="2.0.0" />
+  </ItemGroup>
+<Project>
+```
+
+A project would override the version with the `VersionOverride` metadata on a `<PackageReference />` item:
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>net5.0</TargetFramework>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="PackageA" VersionOverride="3.0.0" />
+  </ItemGroup>
+<Project>
+```
+
+In this case, the project's restore graph would resolve `PackageA` to version `3.0.0`. Any project that references it would also get that version and a user would be responsible for handling an unresolved conflicts.
+
+Finally, a repo owner should be able to disable the ability for developers to override package version. This would be used for instance if someone wanted to ensure that all package versions are unified. This would be possible by setting the MSBuild property `EnablePackageVersionOverride` to `false` in a project or import like `Directory.Build.props`:
+
+```xml
+<Project>
+  <PropertyGroup>
+    <EnablePackageVersionOverride>false</EnablePackageVersionOverride>
+  </PropertyGroup>
+<Project>
+```
+
+When this is disabled, specifying a `VersionOverride` on a `<PackageReference />` would result in a restore error indicating that the feature is disabled.
+
+### Technical explanation
+
+<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->
+
+## Drawbacks
+
```

Comment:
One drawback not cited: this proposed method requires an explicit PackageReference.  Contrast with the `<PackageVersion Update="" />` method, where the version is overridden only if the PackageReference exists in the graph.  This difference becomes evident with a transitive dependency: in the scenario `project -> A -> B`, if I want to override the version of B *only if it is in this project's graph*, the proposed method *forces me to include it in the graph*.  If a future version of A removes that dependency, then my override (per the proposal) would still have B in the graph.

Reply:
That's a very good point. 

I know relying on the Update is super difficult for tooling, so I wonder if this should be a known limitaiton of this design.

We should explicitly ask for feedback from the people that rely on transitive pinning, to gauge their thoughts on this limitation. 

I'm really not sure there's any other solution that doesn't rely on the Update functionality.

Reply:
> I'm really not sure there's any other solution that doesn't rely on the Update functionality.

A possible alternative would be to introduce a `<PackageVersionOverride>` item that could be specified in the project but would only be applied if a matching package existed in the restore graph.  But at that point, it has the complexity of the `<PackageVersion Update="" />` (i.e. I need both a `<PackageReference>` and a `<PackageVersionOverride>` to override a primary reference), or it adds complexity to the overall feature design (i.e. implementing both VersionOverride as metadata and PackageVersionOverride as a separate item).

I'm not sure if that's worth the it vs. keeping the simpler VersionOverride metadata.  Maybe documenting both the VersionOverride and the Update method is sufficient, calling out the tradeoffs for each?

Reply:
Thanks I'll update the proposal to call this out.  I'm personally hoping that VersionOverride is a last resort and therefore a rarity.  Ideally a repo would be able to unify on a single version.  

As discussed here, `<PackageVersion Update="" />` is a viable option and we won't be doing anything to stop that pattern.  We won't be able to detect if these are used incorrectly since its transparent to NuGet that they ran.  I think we'll even place this pattern in the docs for even more advanced use cases.  Just be aware that built-in tooling will respect them when restoring but not always when updating them in a UI or other end-user experience (which is the case with any build customization).